x-pack/plugins/fleet/server/services/agent_policies/package_policies_to_agent_permissions.ts

```diff
@@ -142,22 +144,30 @@ interface DataStreamMeta {
   hidden?: boolean;
   elasticsearch?: {
     privileges?: RegistryDataStreamPrivileges;
+    dynamic_namespace?: boolean;
+    dynamic_dataset?: boolean;
   };
 }
 
 export function getDataStreamPrivileges(dataStream: DataStreamMeta, namespace: string = '*') {
-  let index = `${dataStream.type}-${dataStream.dataset}`;
-
-  if (dataStream.dataset_is_prefix) {
-    index = `${index}.*`;
+  let index = dataStream.hidden ? `.${dataStream.type}-` : `${dataStream.type}-`;
+
+  // Determine dataset
+  if (dataStream.elasticsearch?.dynamic_dataset) {
+    index += `*`;
```

Comment:
I generally find the `dynamic_dataset` setting a little odd because it has to be defined on a data stream level (which has a dataset), but then we allow the collector/input to send to any dataset. It might make more sense to move this setting to the overall package manifest? @jsoriano any thoughts?

Reply:
Yes it was not working for `elasticsearch.*` properties, but I have now fixed this and added tests.

Reply:
Bumping this. If we want to move this to a package-level setting instead of a datastream-level setting we should probably make that happen in this PR. 

> How would it behave at the package level? Would it add the wildcard for all data streams? It could make sense if we want that.

I think we'd have no other choice if this were a top-level package setting than to apply the wildcard to all datastreams for the package, correct?

Reply:
I think @jsoriano's point make sense that we wouldn't necessarily want it for all data streams. You may only need it for a specific type and there's no type specified at the package level. Since the current design still allows for every use case we intend to support and makes it more explicit to package devs that they're only enabling this for a single type at a time, I lean towards keeping this implementation.